### PR TITLE
ui: refac manager send

### PR DESCRIPTION
**Pull request by edouardparis**
based on  #50

### CODE REVIEW

File: src/ui/message.rs [hunk 1]
```diff
@@ -28,6 +31,11 @@ pub enum Message {
     Recipient(usize, RecipientMessage),
     Input(usize, InputMessage),
     AddRecipient,
+    Feerate(u8),
+    GenerateTransaction,
+    SpendTransaction(Result<SpendTransaction, RevaultDError>),
+    Sign(SignMessage),
+    Signed(Result<(), RevaultDError>),
```
Comment (edouardparis) [inline src/ui/message.rs:38]: I will move it to SpendTxMessage

File: src/revaultd/mod.rs [hunk 2]
```diff
@@ -161,6 +163,28 @@ impl RevaultD {
...
+            Some(vec![json!(inputs), json!(outputs), json!(feerate)]),
+        )
+        .map(|mut res: SpendTransaction| {
+            res.feerate = *feerate;
+            res
+        })
+    }
+
+    pub fn update_spend_tx(&self, psbt: &Psbt) -> Result<(), RevaultDError> {
+        let spend_tx = base64::encode(&consensus::serialize(psbt));
```
Comment (darosior) [inline src/revaultd/mod.rs:184]: Small nit: you should use our types from revault_t for the PSBT you know are always going to be of a specific types. It's a nice way to sanity check the tx and to have some more state assumptions that hold for it, should you ever need it

File: src/revaultd/model.rs [hunk 3]
```diff
@@ -167,6 +167,12 @@ pub struct UnvaultTransaction {
     pub unvault_tx: PartiallySignedTransaction,
 }
 
+#[derive(Debug, Clone, Deserialize)]
+pub struct SpendTransaction {
+    #[serde(with = "bitcoin_psbt")]
+    pub spend_tx: PartiallySignedTransaction,
+}
+
```
Comment (darosior) [inline src/revaultd/model.rs]: But sir i already implemented this in revault_tx :(

File: src/revaultd/mod.rs [hunk 4]
```diff
@@ -161,6 +163,28 @@ impl RevaultD {
...
+        inputs: &[String],
+        outputs: &HashMap<String, u64>,
+        feerate: &u8,
+    ) -> Result<SpendTransaction, RevaultDError> {
+        self.call(
+            "getspendtx",
+            Some(vec![json!(inputs), json!(outputs), json!(feerate)]),
+        )
+        .map(|mut res: SpendTransaction| {
+            res.feerate = *feerate;
```
Comment (darosior) [inline src/revaultd/mod.rs:178]: But this field does not exist?

Comment (darosior) [reply src/revaultd/mod.rs:178]: Ok it's added in the next commit..

File: src/ui/view/manager.rs [hunk 5]
```diff
@@ -336,25 +323,249 @@ impl ManagerSelectFeeView {
...
+                                .push(Container::new(
+                                    Row::new()
+                                        .push(text::bold(text::simple(&format!("{}", feerate))))
+                                        .push(text::simple("s/byte")),
+                                ))
+                                .push(
+                                    Container::new(
+                                        Slider::new(
+                                            &mut self.slider,
+                                            0..=200,
```
Comment (darosior) [inline src/ui/view/manager.rs:420]: Should start at `1` (credits to Ed)

File: src/ui/state/manager.rs [hunk 6]
```diff
@@ -151,25 +160,44 @@ impl From<ManagerHomeState> for Box<dyn State> {
...
 impl ManagerSendState {
     pub fn new(revaultd: Arc<RevaultD>) -> Self {
         ManagerSendState {
             revaultd,
-            view: ManagerSendView::new(),
+            step: ManagerSendStep::SelectOutputs(ManagerSelectOutputsView::new()),
             warning: None,
             vaults: Vec::new(),
             outputs: vec![ManagerSendOutput::new()],
+            feerate: 20,
```
Comment (darosior) [inline src/ui/state/manager.rs:198]: Why is it hardcoded to 20 ? Also it seems to  be part of the `psbt` tuple as well

File: src/ui/state/manager.rs [hunk 7]
```diff
@@ -199,27 +227,134 @@ impl ManagerSendState {
...
-            Message::Previous => self.view = self.view.previous(),
+            Message::Signed(res) => match res {
+                Ok(_) => {
+                    if let ManagerSendStep::Sign { signer, .. } = &mut self.step {
+                        // During this step state has a generated psbt
+                        // and signer has a signed psbt.
+                        self.psbt = Some((
+                            signer.signed_psbt.clone().unwrap(),
+                            self.psbt.clone().unwrap().1,
+                        ));
```
Comment (darosior) [inline src/ui/state/manager.rs:295]: Could you expand a bit on why the signer always has a signed PSBT and it's safe to unwrap there?

**Review by darosior — COMMENTED** [review]
Just a first round of comment after skimming through it

File: src/ui/view/manager.rs [hunk 8]
```diff
@@ -336,25 +323,249 @@ impl ManagerSelectFeeView {
...
+                Container::new(text::bold(text::simple("Generated transaction")))
+                    .width(Length::Fill)
+                    .align_x(Align::Center),
+            )
+            .push(
+                Column::new()
+                    .push(
+                        Row::new()
+                            .push(text::simple("Feerate: "))
+                            .push(text::bold(text::simple(&format!("{} s/byte", feerate)))),
```
Comment (danielabrozzoni) [inline src/ui/view/manager.rs:510]: nit: I would use `sats/vbyte`? What do you think?

Comment (edouardparis) [reply src/ui/view/manager.rs:510]: good idea

**Comment by danielabrozzoni** [comment]
The copy-paste PSBT button here doesn't work:

Revault GUI_006

File: src/ui/state/manager.rs [hunk 9]
```diff
@@ -199,27 +227,134 @@ impl ManagerSendState {
...
+            .collect()
+    }
 }
 
 impl State for ManagerSendState {
     fn update(&mut self, message: Message) -> Command<Message> {
         match message {
+            Message::SpendTransaction(res) => {
+                self.processing = false;
+                match res {
```
Comment (danielabrozzoni) [inline src/ui/state/manager.rs:251]: small nit: you can use `if let Ok(tx) = res`

File: src/ui/state/manager.rs [hunk 10]
```diff
@@ -199,27 +227,134 @@ impl ManagerSendState {
...
+                        self.psbt = Some((tx.spend_tx, tx.feerate));
+                    }
+                    Err(e) => self.warning = Some(Error::RevaultDError(e)),
+                }
+            }
+            Message::GenerateTransaction => {
+                self.processing = true;
+                let inputs = self
+                    .selected_inputs()
+                    .iter()
```
Comment (danielabrozzoni) [inline src/ui/state/manager.rs:262]: I think you should `into_iter()` here

File: src/ui/state/manager.rs (hunk 7 above)
Comment (darosior) [reply src/ui/state/manager.rs:295]: :eyes:

**Comment by darosior** [comment]
Feedback from testing:
- looking really good UI wise
- i think the flow is right (choose your outputs, then your inputs)
- needs to do some best effort fee computation to prevent the user from paying too much and not choosing enough vaults to pay for 1. the feerate they ask 2. the under-the-hood added CPFP output
- needs to output the error on transaction generation error (maybe the raw JSONRPC error from revaultd for now, maybe do better in the future)

File: src/ui/view/manager.rs [hunk 11]
```diff
@@ -336,25 +323,249 @@ impl ManagerSelectFeeView {
...
+                    .align_x(Align::Center),
+            )
+            .push(Container::new(
+                Row::new()
+                    .push(text::bold(text::simple(&format!("{}", feerate))))
+                    .push(text::simple("satoshis/vbyte")),
+            ))
+            .push(
+                Container::new(
+                    Slider::new(&mut self.slider, 1..=1000, *feerate, Message::Feerate)
```
Comment (darosior) [inline src/ui/view/manager.rs:390]: Would be nice to be able to input it too in the future :tm:

**Review by darosior — APPROVED** [review]
ACK 09d98f827cf13daf8fd74cdfea46374add38a285

Would be nice to:
- Have a title to the coin selection screen, something like "Select the vaults you want to spend"
- Be able to select a vault by just clicking on the row

**Pull request merged**